File: Hard/1547-minimum-cost-to-cut-a-stick.cpp
#include <iostream>

using namespace std;
#include <vector>

class Solution {
public:
  #define ll long long
  ll dp[102][102];
  int minCost(int n, vector<int>& cuts) {
    sort(cuts.begin(), cuts.end());
    cuts.insert(cuts.begin(), 0);
    cuts.push_back(n);
    for (int i = cuts.size()-1; ~i; --i) {
      for (int j = i + 1; j<cuts.size(); ++j) {
        if (i+1<j) {
          dp[i][j] = LLONG_MAX;
          for (int k = i + 1; k < j; ++k) {
            dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j]);
          }
          dp[i][j] += cuts[j] - cuts[i];
        }
        cout << i << " " << j << " " << dp[i][j] << " \n";
      }
    }
    return dp[0][cuts.size()-1];
  }
};

int main()
{
  Solution sol;
  int n = 9;
  vector<int> cuts;
  cuts.push_back(5);
  cuts.push_back(6);
  cuts.push_back(1);
  cuts.push_back(4);
  cuts.push_back(2);
  int ans = sol.minCost(n, cuts);
  cout << "answer is: " << ans << endl;
}

File: Hard/1547-minimum-cost-to-cut-a-stick.py
from typing import List
import sys

class Solution:
  def minCost(self, n: int, cuts: List[int]) -> int:
    def dp(start, end):
      print('start: ', start, 'end: ', end)
      if (start, end) in memo:
        return memo[(start, end)]

      ans = sys.maxsize
      canCut = False
      print('cuts: ', cuts)
      for cut in cuts:
        if start < cut < end:
          canCut = True 
          ans = min(ans, dp(start, cut) + dp(cut, end) + end - start)
      if not canCut:
        return 0

      memo[(start, end)] = ans

      return ans
    memo = {}
    return dp(0, n)

sol = Solution()
ans = sol.minCost(7, [1, 3, 4, 5])
print(ans)